### PR TITLE
Refactor train!

### PR DESCRIPTION
**Pull request by dhairyagandhi96**
This is in response to #666 where we can visualise `train!` as a wrapper around the `step!` function, while maintaining the same api.

To actually accumulate the loss as in https://github.com/FluxML/Flux.jl/issues/666#issuecomment-471307139, we could of course go with the loss function being a simple closure here, getting rid of the need to send around the mini batch.

With the changes to Zygote, it might be nice to actually have this in our interface.

cc @MikeInnes @oxinabox

### CODE REVIEW

File: src/optimise/train.jl [hunk 1]
```diff
@@ -42,6 +42,13 @@ function stop()
   throw(StopException())
 end
 
+function step!(loss, ps, minibatch, opt)
```
Comment (oxinabox) [inline src/optimise/train.jl]: Can this have a docstring, and be exported?

Comment (dhairyagandhi96) [reply src/optimise/train.jl]: Yeah, it would be good to converge on its implementation, with that I'll add the doctrings and a basic test too.

File: src/optimise/train.jl [hunk 2]
```diff
@@ -42,6 +42,13 @@ function stop()
...
+  gs = gradient(ps) do
+    loss(minibatch...)
+  end
+  update!(opt, ps, gs)
```
Comment (CarloLucibello) [inline src/optimise/train.jl]: Isn't this return incompatible with the interface proposed in https://github.com/FluxML/Flux.jl/issues/666#issuecomment-516240099?

Comment (dhairyagandhi96) [reply src/optimise/train.jl]: It returns `nothing`, which should be fine for most cases. For that comment specifically, yes, but it really is just a matter of agreeing on the semantics desired out of it. returning the loss here should be fine, but seeing as it is stepping through a batch, and doing a training step, I feel the correct return should be a nothing, with a slightly more trained model as it were

**Comment by dhairyagandhi96** [comment]
Pinging @MikeInnes for his thoughts on the semantics that would make most sense here.

**Comment by MikeInnes** [comment]
I agree that closing over the batch would make the most sense. I also agree it's best to avoid returning things like losses, since as discussed earlier we can just close over those things, and we will eventually want an out-of-place version of this that returns a model.

`train!` will likely need a fair bit of refactoring once we figure out new optimisers + accelerator support, but we can figure that out later.
